src/Tools/Microsoft.dotnet-openapi/src/Commands/BaseCommand.cs

```diff
@@ -0,0 +1,426 @@
+                    throw new ArgumentException($"The project '{project}' does not exist.");
+                }
+            }
+            else
+            {
+                var projects = Directory.GetFiles(WorkingDirectory, "*.csproj", SearchOption.TopDirectoryOnly);
+                if (projects.Length == 0)
+                {
+                    throw new ArgumentException("No project files were found in the current directory. Either move to a new directory or provide the project explicitly");
+                }
+                if (projects.Length > 1)
+                {
+                    throw new ArgumentException("More than one project was found in this directory, either remove a duplicate or explicitly provide the project.");
+                }
+
+                project = projects[0];
+            }
+
+            return new FileInfo(project);
+        }
+
+        protected Project LoadProject(FileInfo projectFile)
+        {
+            var project = ProjectCollection.GlobalProjectCollection.LoadProject(
+                projectFile.FullName,
+                globalProperties: null,
+                toolsVersion: null);
+            project.ReevaluateIfNecessary();
+            return project;
+        }
+
+        internal bool IsProjectFile(string file)
+        {
+            return File.Exists(Path.GetFullPath(file)) && file.EndsWith(".csproj");
+        }
+
+        internal bool IsUrl(string file)
+        {
+            return Uri.TryCreate(file, UriKind.Absolute, out var _) && file.StartsWith("http");
+        }
+
+        internal void AddServiceReference(
+            string tagName,
+            FileInfo projectFile,
+            string sourceFile,
+            CodeGenerator codeGenerator,
+            string sourceUrl = null)
+        {
+            var project = LoadProject(projectFile);
+            var items = project.GetItems(tagName);
+            var fileItems = items.Where(i => string.Equals(GetFullPath(i.EvaluatedInclude), GetFullPath(sourceFile), StringComparison.Ordinal));
+
+            if (fileItems.Count() >= 1)
+            {
+                Warning.Write($"One or more references to {sourceFile} already exist. Duplicate references could lead to unexpected behavior.");
+                return;
+            }
+
+            if (sourceUrl != null)
+            {
+                if (items.Any(i => string.Equals(i.GetMetadataValue(SourceUrlAttrName), sourceUrl)))
+                {
+                    Warning.Write($"A reference to '{sourceUrl}' already exists in '{project.FullPath}'.");
+                    return;
+                }
+            }
+
+            if (fileItems.Count() == 0)
+            {
+                var metadata = new Dictionary<string, string>();
+
+                if (!string.IsNullOrEmpty(sourceUrl))
+                {
+                    metadata[SourceUrlAttrName] = sourceUrl;
+                }
+
+                metadata[CodeGeneratorAttrName] = codeGenerator.ToString();
```

Comment:
Either change `GetCodeGenerator(...)` to return `null` when the user doesn't specify the option or special-case the default (`NSwagCSharp`) to _not_ write out the metadata in that default case. The first option is better because it allows users to be explicit about the generator when they wish but worse because it'll mess up your packages lookup. Maybe a new `GetCodeGeneratorOrDefault(...)` method for packages lookup?

Reply:
Why is it important that we not have the CodeGenerator specified in the xml when it was defaulted? I was given to believe that the "default" existed only in this tool anyway, is it the case  that the SDK stuff which generates the code also defaults to CSharp?

Reply:
Users can change the code generator their projects use. When writing out an explicit value here, you're making it more difficult for them to switch. That is, they need to manually edit the project file and not just change the referenced packages.